### ISSUE TITLE
(Again) How did you arrive to the normalization constants?

### ISSUE THREAD

**Issue by manolotis** [stepankonev/waymo-motion-prediction-challenge-2022-multipath-plus-plus#23]
First of all, thank you for making this implementation publicly available. I find your code really elegant. 

I have some questions regarding how you arrived to the normalization constants. I already saw a really similar issue (#1), but it did not completely clarify what I am wondering. 

I see that normalization is now performed in `model.data.py`, with these means and standard deviations:

https://github.com/stepankonev/waymo-motion-prediction-challenge-2022-multipath-plus-plus/blob/4636641b00cd16f62f82a612c728200f3fc4a32b/code/model/data.py#L30-L48

And to the predicted coordinates during training with these (if specified):
https://github.com/stepankonev/waymo-motion-prediction-challenge-2022-multipath-plus-plus/blob/4636641b00cd16f62f82a612c728200f3fc4a32b/code/train.py#L83-L85

For a different model I am developing, I tried to calculate similar constants (mainly for various features of target/history and target/future), and I arrive to considerably different values. For some features, I get fairly similar values, but for other they are a lot higher (a factor of 10-100, especially noticeable in the coordinates).

My approach to compute these values has been to first prerender the dataset using `MultiPathPPRenderer` without normalization and filtering only interesting agents. Then traversing all the prerendered scenarios from the training split and computing the mean and standard deviation of each feature for the target agent. How come I am getting such different values? Could you elaborate on the part of the data you used to compute these constants? In particular: (1) did you use a subset of agents? (e.g. only interesting, or fully observed), (2) did you use a subset of the scenarios? 

Thank you in advance!

**Comment by zheningy** [comment]
Hi @manolotis, thanks for pointing this out! I have similar questions about this. Btw how do you calculate the constants for **target/future/xy**? I do not quite understand why we need this constant.

**Comment by manolotis** [comment]
> Hi @manolotis, thanks for pointing this out! I have similar questions about this. Btw how do you calculate the constants for **target/future/xy**? I do not quite understand why we need this constant.

I calculate the constants for target/future/xy in a similar fashion as for target/history/xy:  basically aggregating the mean and std for the future coordinates (ofc without counting the invalid timestamps).

Generally it's beneficial to normalize the outputs if they have a significantly different magnitude. Here I suppose it's beneficial given the differences in resulting future x and y (since the inputs are processed such that the target is always facing the positive x direction at prediction time).

I experimented not normalizing the inputs as suggested by https://github.com/stepankonev/waymo-motion-prediction-challenge-2022-multipath-plus-plus/issues/3#issuecomment-1174631507, but I always run into issues with infinite values in the first epoch, and when changing the `running_mean_mode` as suggested by https://github.com/stepankonev/waymo-motion-prediction-challenge-2022-multipath-plus-plus/issues/3#issuecomment-1181579038 the same issues appear after a few epochs. After a little digging, the problem arises from `torch.logdet` of the covariances during the loss computation, since the determinant ends up negative at some point. I will continue with the normalized output, but I am still curious how the normalization constants were computed exactly.

**Comment by zheningy** [comment]
Thanks for your explanation! I meet the same issue when setting the `normalize_output` as False.
